hrm-service/src/permissions/index.ts

```diff
@@ -4,70 +4,39 @@ export { rulesMap } from './rulesMap';
 export { Actions, actionsMaps, getActions } from './actions';
 
 import { setupCanForRules } from './setupCanForRules';
-import { rulesMap } from './rulesMap';
+import { RulesFile } from './rulesMap';
 // eslint-disable-next-line prettier/prettier
 import type { Request, Response, NextFunction } from 'express';
 
-const initializedCanMap = Object.entries(rulesMap).reduce<
-  { [key in keyof typeof rulesMap]: ReturnType<typeof setupCanForRules> }
->((accum, [key, rules]) => {
-const can = setupCanForRules(rules);
-return {
-  ...accum,
-  [key]: can,
+const canCache: Record<string, ReturnType<typeof setupCanForRules>> = {};
+
+export type Permissions = {
+  rules: (accountSid: string) => RulesFile;
+  cachePermissions: boolean;
 };
-}, null);
 
 /**
  * Applies the permissions if valid.
  * @throws Will throw if initializedCan is not a function
  */
 export const applyPermissions = (req: Request,
   initializedCan: ReturnType<typeof setupCanForRules>,
-  permissionsConfigName: string,
   ) => {
-  if (!initializedCan) throw new Error(`Cannot find rules for ${permissionsConfigName}`);
-
-  if (typeof initializedCan === 'string')
-    throw new Error(`Error in rules for ${permissionsConfigName}. Error: ${initializedCan}`);
-
   if (typeof initializedCan !== 'function')
-    throw new Error(`Error in rules for ${permissionsConfigName}. Error: can is not a function.`);
+    throw new Error(`Error in looked up permission rules: can is not a function.`);
 
   //@ts-ignore TODO: Improve our custom Request type to override Express.Request
   req.can = initializedCan;
 };
 
-/**
- * @throws Will throw if there is no env var set for PERMISSIONS_${accountSid} or if it's an invalid key in rulesMap
- */
- export const getPermissionsConfigName = (accountSid: string) => {
-  const permissionsKey = `PERMISSIONS_${accountSid}`;
-
-  const permissionsConfigName = process.env[permissionsKey];
-
-  if (!permissionsConfigName)
-    throw new Error(`No permissions set for account ${accountSid}.`);
-
-  if (!rulesMap[permissionsConfigName])
-    throw new Error(`Permissions rules with name ${permissionsConfigName} missing in rules map.`);
-
-  return permissionsConfigName;
-};
-
-
-export const setupPermissions = (req: Request, res: Response, next: NextFunction) => {
-  if (process.env.USE_OPEN_PERMISSIONS) {
-    applyPermissions(req, initializedCanMap.open, 'open rules');
-    return next();
+export const setupPermissions = (lookup: Permissions) => (req: Request, res: Response, next: NextFunction) => {
+  const { accountSid } = <any>req;
+  if (lookup.cachePermissions) {
```

Comment:
Since the new permissions lookup API gets a rules file based on the provided accountSid, a naive implementation  would need to convert the rules to a can function for every request, whereas the previous code converted every possible account's rules up front, which is much more efficient

To retain this efficiency, we now lazily cache the can function created when an accountSid is seen for the first time then reuse it for subsequent calls to the same account.

The option to disable this cache is used in tests, where we want to inject different permission rules for the same account in each test, but don't want to restart the Express server (we could restart the server every test, but this way is faster :-P)